sql/moz-fx-data-shared-prod/subscription_platform/apple_subscriptions/view.sql

```diff
@@ -0,0 +1,225 @@
+    UNNEST(ARRAY<STRUCT<document_id STRING, `timestamp` TIMESTAMP, `data` STRING>>[])
+  UNION ALL
+  SELECT
+    CAST(NULL AS STRING) AS document_id,
+    *
+  FROM
+    `moz-fx-data-shared-prod`.mozilla_vpn_derived.guardian_apple_events_v1
+  WHERE
+    environment = "Production"
+),
+apple_iap_times AS (
+  SELECT
+    original_transaction_id,
+    ARRAY_CONCAT(
+      -- original_purchase_date does not change
+      [MIN(original_purchase_date)],
+      -- status 1 means subscription is active
+      IFNULL(ARRAY_AGG(DISTINCT IF(status IN (1, 3, 4), purchase_date, NULL) IGNORE NULLS), [])
+    ) AS start_times,
+    ARRAY_CONCAT(
+      -- status 2 means subscription is expired
+      IFNULL(ARRAY_AGG(DISTINCT IF(status = 2, expires_date, NULL) IGNORE NULLS), []),
+      -- status 5 means subscription is revoked
+      IFNULL(ARRAY_AGG(DISTINCT IF(status = 5, revocation_date, NULL) IGNORE NULLS), []),
+      -- only use most recent expires_date when not limited to status 2
+      ARRAY_AGG(expires_date ORDER BY verified_at DESC LIMIT 1)
+    ) AS end_times,
+  FROM
+    apple_iap_events
+  GROUP BY
+    original_transaction_id
+),
+apple_iap_periods AS (
+  SELECT
+    original_transaction_id,
+    start_time,
+    end_time,
+    period_offset,
+  FROM
+    apple_iap_times
+  LEFT JOIN
+    UNNEST(
+      ARRAY(
+        SELECT AS STRUCT
+          (
+            SELECT
+              MIN(end_time)
+            FROM
+              UNNEST(end_times) AS end_time
+            WHERE
+              end_time > start_time
+          ) AS end_time,
+          MIN(start_time) AS start_time,
+        FROM
+          UNNEST(start_times) AS start_time
+        GROUP BY
+          end_time
+        ORDER BY
+          start_time
+      )
+    )
+    WITH OFFSET AS period_offset
+),
+apple_iap_period_aggregates AS (
+  SELECT
+    periods.original_transaction_id,
+    periods.start_time,
+    periods.end_time,
+    periods.period_offset,
+    COALESCE(
+      MAX(events.legacy_subscription_id),
+      periods.original_transaction_id
+    ) AS original_subscription_id,
+    MIN(events.original_purchase_date) AS original_purchase_date,
+    MAX(events.user_id) AS fxa_uid,
+    ARRAY_AGG(DISTINCT events.offer_identifier IGNORE NULLS) AS promotion_codes,
+    ARRAY_AGG(
+      STRUCT(
+        events.auto_renew_status,
+        events.bundle_id,
+        events.expiration_intent,
+        events.grace_period_expires_date,
+        events.original_transaction_id,
+        events.product_id,
+        events.revocation_reason,
+        events.status,
+        events.verified_at
+      )
+      ORDER BY
+        events.verified_at DESC
+      LIMIT
+        1
+    )[OFFSET(0)].*,
+  FROM
+    apple_iap_periods AS periods
+  LEFT JOIN
+    apple_iap_events AS events
+  ON
+    periods.original_transaction_id = events.original_transaction_id
+    AND periods.start_time <= events.purchase_date
+    AND periods.end_time > events.purchase_date
```

Comment:
I think this solves the issue of aggregating events from the wrong period, but I'm not sure if the join conditions are sufficient

Reply:
I think this works correctly with how the `apple_iap_periods` CTE calculates the periods.

I do wonder if an inner join might be more appropriate though.

Reply:
sure